File: pubbot/pubbot.py
import requests
import xml.etree.ElementTree as ET
import json
from datetime import datetime

def fetch_pubmed_publications(term="Giacomo B Marino", page=2):
    # Pagination math
    retstart = 1
    retmax = 50

    # Step 1: ESearch to get PMIDs
    esearch_url = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi"
    params = {
        "db": "pubmed",
        "term": term,
        "retstart": retstart,
        "retmax": retmax,
        "sort": "pub+date",
        "retmode": "json"
    }
    esearch_response = requests.get(esearch_url, params=params).json()
    pmids = esearch_response["esearchresult"]["idlist"]

    if not pmids:
        return {"publications": []}

    # Step 2: EFetch to get details
    efetch_url = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi"
    efetch_params = {
        "db": "pubmed",
        "id": ",".join(pmids),
        "retmode": "xml"
    }
    efetch_response = requests.get(efetch_url, params=efetch_params)
    root = ET.fromstring(efetch_response.content)

    publications = []
    for article in root.findall(".//PubmedArticle"):
        try:
            article_data = article.find("MedlineCitation/Article")
            title = article_data.findtext("ArticleTitle")
            journal = article_data.findtext("Journal/Title")
            pub_date_elem = article_data.find("Journal/JournalIssue/PubDate")
            date_parts = [pub_date_elem.findtext(tag) for tag in ("Year", "Month", "Day")]
            date = "-".join(filter(None, date_parts))
            if len(date_parts) == 1:
                date += "-01-01"

            authors = article_data.find("AuthorList")
            contributors = []
            for author in authors.findall("Author"):
                last = author.findtext("LastName")
                initials = author.findtext("Initials")
                if last and initials:
                    contributors.append(f"{last} {initials}")

            author_str = next((a for a in contributors if "Marino" in a), contributors[0] if contributors else "Unknown")

            pub_type = article_data.findtext("PublicationTypeList/PublicationType")
            article_ids = article.find("PubmedData/ArticleIdList")

            doi = pmid = pmc = None
            for id_elem in article_ids.findall("ArticleId"):
                id_type = id_elem.attrib.get("IdType")
                if id_type == "doi":
                    doi = id_elem.text
                elif id_type == "pubmed":
                    pmid = id_elem.text
                elif id_type == "pmc":
                    pmc = id_elem.text

            publications.append({
                "title": title,
                "journal": journal,
                "date": date,
                "type": pub_type,
                "author": author_str,
                "doi": doi,
                "pmid": pmid,
                "pmc": pmc,
                "contributors": contributors,
                "source": "Giacomo Marino",
                "grade": "Preferred source (of 2)"
            })

        except Exception as e:
            print("Error parsing article:", e)
            continue

    return {"publications": publications}


# Example usage:
pubs = fetch_pubmed_publications()

with open('src/app/publications.json', 'r') as f:
    data = json.load(f)

    # Combine publications from both sources, using DOI as unique identifier
    existing_dois = {pub["doi"] for pub in data["publications"] if pub["doi"]}
    
    # Only add publications that don't already exist
    for pub in pubs["publications"]:
        if pub["doi"] not in existing_dois:
            data["publications"].append(pub)
            existing_dois.add(pub["doi"])
    
    # Sort publications by date, most recent first
    def parse_date(date_str):
        try:
            return datetime.strptime(date_str, "%Y-%m-%d")
        except ValueError:
            try:
                return datetime.strptime(date_str, "%Y-%b-%d")
            except ValueError:
                # Handle dates with only year and month
                return datetime.strptime(date_str, "%Y-%b")
    
    # Remove bioRxiv versions if a published version exists
    title_to_pub = {}
    for pub in data["publications"]:
        title = pub["title"].lower().replace(".", "")  # Normalize title for comparison
        if title not in title_to_pub:
            title_to_pub[title] = pub
        else:
            # If we find a duplicate, keep the non-bioRxiv version
            existing_pub = title_to_pub[title]
            current_is_biorxiv = "biorxiv" in pub.get("journal", "").lower()
            existing_is_biorxiv = "biorxiv" in existing_pub.get("journal", "").lower()
            
            # Keep the non-biorxiv version if one exists
            if current_is_biorxiv and not existing_is_biorxiv:
                continue  # Skip this biorxiv version
            elif not current_is_biorxiv and existing_is_biorxiv:
                title_to_pub[title] = pub  # Replace biorxiv with published version
            else:
                # If both are biorxiv or both are not biorxiv, keep the most recent
                if parse_date(pub["date"]) > parse_date(existing_pub["date"]):
                    title_to_pub[title] = pub
    
    # Update publications list with deduplicated entries
    data["publications"] = list(title_to_pub.values())
    
    # Sort the deduplicated list
    data["publications"].sort(key=lambda x: parse_date(x["date"]), reverse=True)

print(len(pubs["publications"]))

with open('src/app/publications.json', 'w') as f:
    json.dump(data, f, indent=4)

File: src/lib/pubbot.ts
export interface Publication {
  title: string;
  journal: string;
  date: string;
  type: string;
  author: string;
  doi: string | null;
  pmid: string | null;
  pmc: string | null;
  contributors: string[];
  source: string;
  grade: string;
}

export interface PublicationsData {
  publications: Publication[];
}

interface ESummaryArticle {
  title?: string;
  fulljournalname?: string;
  pubdate?: string;
  pubtype?: string[];
  authors?: { name: string }[];
  articleids?: { idtype: string; value: string }[];
}

function formatEsummaryDate(pubdate: string): string {
  const match = pubdate.trim().match(/^(\d{4})\s+(\w{3})(?:\s+(\d{1,2}))?/);
  if (!match) return pubdate;

  const [, year, month, day] = match;
  if (day) return `${year}-${month}-${day.padStart(2, "0")}`;
  return `${year}-${month}`;
}

export function parsePublicationDate(dateStr: string): Date {
  const formats = [
    /^(\d{4})-(\d{2})-(\d{2})$/,
    /^(\d{4})-(\w{3})-(\d{2})$/,
    /^(\d{4})-(\w{3})$/,
    /^(\d{4})$/,
  ];

  const monthMap: Record<string, number> = {
    Jan: 0,
    Feb: 1,
    Mar: 2,
    Apr: 3,
    May: 4,
    Jun: 5,
    Jul: 6,
    Aug: 7,
    Sep: 8,
    Oct: 9,
    Nov: 10,
    Dec: 11,
  };

  const isoMatch = dateStr.match(formats[0]);
  if (isoMatch) {
    return new Date(Number(isoMatch[1]), Number(isoMatch[2]) - 1, Number(isoMatch[3]));
  }

  const fullMatch = dateStr.match(formats[1]);
  if (fullMatch) {
    return new Date(Number(fullMatch[1]), monthMap[fullMatch[2]], Number(fullMatch[3]));
  }

  const monthMatch = dateStr.match(formats[2]);
  if (monthMatch) {
    return new Date(Number(monthMatch[1]), monthMap[monthMatch[2]], 1);
  }

  const yearMatch = dateStr.match(formats[3]);
  if (yearMatch) {
    return new Date(Number(yearMatch[1]), 0, 1);
  }

  return new Date(0);
}

export async function fetchPubmedPublications(
  term = "Giacomo B Marino"
): Promise<PublicationsData> {
  const esearchUrl = new URL(
    "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi"
  );
  esearchUrl.searchParams.set("db", "pubmed");
  esearchUrl.searchParams.set("term", term);
  esearchUrl.searchParams.set("retstart", "0");
  esearchUrl.searchParams.set("retmax", "50");
  esearchUrl.searchParams.set("sort", "pub+date");
  esearchUrl.searchParams.set("retmode", "json");

  const esearchResponse = await fetch(esearchUrl);
  if (!esearchResponse.ok) {
    throw new Error(`PubMed esearch failed: ${esearchResponse.status}`);
  }

  const esearchJson = await esearchResponse.json();
  const pmids: string[] = esearchJson?.esearchresult?.idlist ?? [];

  if (pmids.length === 0) {
    return { publications: [] };
  }

  const esummaryUrl = new URL(
    "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi"
  );
  esummaryUrl.searchParams.set("db", "pubmed");
  esummaryUrl.searchParams.set("id", pmids.join(","));
  esummaryUrl.searchParams.set("retmode", "json");

  const esummaryResponse = await fetch(esummaryUrl);
  if (!esummaryResponse.ok) {
    throw new Error(`PubMed esummary failed: ${esummaryResponse.status}`);
  }

  const esummaryJson = await esummaryResponse.json();
  const publications: Publication[] = [];

  for (const pmid of pmids) {
    const article = esummaryJson?.result?.[pmid] as ESummaryArticle | undefined;
    if (!article?.title) continue;

    const contributors = (article.authors ?? [])
      .map((author) => author.name)
      .filter(Boolean);

    const author =
      contributors.find((name) => name.includes("Marino")) ??
      contributors[0] ??
      "Unknown";

    let doi: string | null = null;
    let pmc: string | null = null;
    for (const id of article.articleids ?? []) {
      if (id.idtype === "doi") doi = id.value;
      if (id.idtype === "pmc") pmc = id.value;
    }

    publications.push({
      title: article.title,
      journal: article.fulljournalname ?? "",
      date: article.pubdate ? formatEsummaryDate(article.pubdate) : "",
      type: article.pubtype?.[0] ?? "Journal Article",
      author,
      doi,
      pmid,
      pmc,
      contributors,
      source: "Giacomo Marino",
      grade: "Preferred source (of 2)",
    });
  }

  return { publications };
}

export function mergePublications(
  existing: PublicationsData,
  incoming: PublicationsData
): PublicationsData {
  const data: PublicationsData = {
    publications: [...existing.publications],
  };

  const existingDois = new Set(
    data.publications.map((pub) => pub.doi).filter((doi): doi is string => Boolean(doi))
  );

  for (const pub of incoming.publications) {
    if (pub.doi && !existingDois.has(pub.doi)) {
      data.publications.push(pub);
      existingDois.add(pub.doi);
    }
  }

  const titleToPub = new Map<string, Publication>();
  for (const pub of data.publications) {
    const title = pub.title.toLowerCase().replace(/\./g, "");
    const existingPub = titleToPub.get(title);

    if (!existingPub) {
      titleToPub.set(title, pub);
      continue;
    }

    const currentIsBiorxiv = pub.journal.toLowerCase().includes("biorxiv");
    const existingIsBiorxiv = existingPub.journal.toLowerCase().includes("biorxiv");

    if (currentIsBiorxiv && !existingIsBiorxiv) {
      continue;
    }

    if (!currentIsBiorxiv && existingIsBiorxiv) {
      titleToPub.set(title, pub);
      continue;
    }

    if (parsePublicationDate(pub.date) > parsePublicationDate(existingPub.date)) {
      titleToPub.set(title, pub);
    }
  }

  const publications = Array.from(titleToPub.values()).sort(
    (a, b) => parsePublicationDate(b.date).getTime() - parsePublicationDate(a.date).getTime()
  );

  return { publications };
}

export function serializePublications(data: PublicationsData): string {
  return `${JSON.stringify(data, null, 4)}\n`;
}
